corehq/apps/auditcare/tasks.py

```diff
@@ -0,0 +1,86 @@
+import logging
+import sys
+
+from celery.schedules import crontab
+from celery.task import periodic_task
+from couchdbkit.ext.django.loading import get_db
+
+from dimagi.utils.couch.database import iter_docs
+from dimagi.utils.dates import force_to_datetime
+from dimagi.utils.logging import notify_exception
+
+from corehq.apps.auditcare.models import (
+    ACCESS_LOGIN,
+    ACCESS_LOGOUT,
+    ACCESS_FAILED,
+    AccessAudit,
+    NavigationEventAudit,
+)
+from corehq.apps.auditcare.utils.export import get_sql_start_date
+from corehq.util.soft_assert import soft_assert
+
+log = logging.getLogger(__name__)
+
+ACCESS_LOOKUP = {
+    "login": ACCESS_LOGIN,
+    "logout": ACCESS_LOGOUT,
+    "login_failed": ACCESS_FAILED,
+}
+
+
+@periodic_task(run_every=crontab(minute="30"), queue='background_queue')
+def copy_events_to_sql(limit=1000):
+    db = get_db("auditcare")
+    start_date = get_sql_start_date()
+    log.info(f"Initial start date: {start_date}")
+
+    startkey = [
+        start_date.year,
+        start_date.month,
+        start_date.day,
+        start_date.hour,
+        start_date.minute,
+        start_date.second,
+        start_date.microsecond,
+    ]
+    ids = [r["id"] for r in db.view(
+        "auditcare/all_events",
+        startkey=startkey,
+        reduce=False,
+        include_docs=False,
+        descending=True,
+        limit=limit,
+    )]
+    for doc in iter_docs(db, ids, chunksize=1000):
+        try:
+            kwargs = _pick(doc, ["user", "domain", "ip_address", "session_key",
+                                 "headers", "status_code", "user_agent"])
+            kwargs.update({
+                "event_date": force_to_datetime(doc.get("event_date")),
+            })
+            if doc["doc_type"] == "NavigationEventAudit":
+                kwargs.update(_pick(doc, ["params", "headers", "status_code", "view", "view_kwargs"]))
+                kwargs.update({
+                    "path": doc.get("request_path"),
+                })
+                NavigationEventAudit(**kwargs).save()
+            elif doc["doc_type"] == "AccessAudit":
+                kwargs.update(_pick(doc, ["http_accept", "trace_id"]))
+                kwargs.update({
+                    "access_type": ACCESS_LOOKUP.get(doc.get("access_type")),
+                    "path": doc.get("path_info"),
+                })
+                AccessAudit(**kwargs).save()
```

Comment:
Would be more efficient (faster and incur less load on the database) to accumulate a batch and then use `AccessAudit.objects.bulk_create(objects_to_save)` (same applies to `NavigationEventAudit` above).

Reply:
I took a stab at this but realized if one bulk save succeeds and the next fails, it would either introduce gaps or I'd need to add rollback logic for the save that succeeded. I'd rather run this for a little while and have more confidence it won't error out, then add bulk saving to speed it up.